File: app/src/main/java/com/goel/attendancetracker/OrganisationActivity.java
package com.goel.attendancetracker;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.ContentValues;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.res.ColorStateList;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.graphics.Typeface;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.core.content.res.ResourcesCompat;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.goel.attendancetracker.classes.ClassesAdapter;
import com.goel.attendancetracker.classes.ClassesModel;
import com.goel.attendancetracker.database.DatabaseHandler;
import com.goel.attendancetracker.database.Params;
import com.goel.attendancetracker.downloadmanager.ClassDownloadManager;
import com.goel.attendancetracker.downloadmanager.FileDataModel;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Objects;

public class OrganisationActivity extends AppCompatActivity implements EditDialogBox.EditDialogListener, AddDialogBox.AddDialogListener, MarkDialogBox.MarkAttendanceActions {


    private static final int STORAGE_PERMISSION_TOKEN = 1;
    private String organisationName;
    private String organisationId;
    private DatabaseHandler databaseHandler;
    private ProgressBar overallProgress;
    private int overallAttendance;
    private int overallRequiredAttendance;

    private RecyclerView classContainer;
    private ClassesAdapter classAdapter;
    private ArrayList<ClassesModel> classList;
    private ClassesModel focusedClass;


    @SuppressLint("SetTextI18n")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_organisation);

        organisationId = Params.OPEN_ORG;

        classContainer = findViewById(R.id.class_grid);
        classList = new ArrayList<>();

        databaseHandler = new DatabaseHandler(OrganisationActivity.this);

        initializeOrganisation();
        Objects.requireNonNull(getSupportActionBar()).setTitle(organisationName.toUpperCase());

        setAdapter();
        getClassList();
        refreshProgress();

        ((TextView)findViewById(R.id.target_value)).setText(overallRequiredAttendance + "%");

        setClickListeners();
    }

    //-- Context Menu
    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.organisation_toolbar_menu, menu);
        return true;
    }

    @SuppressLint("NonConstantResourceId")
    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()){
            case R.id.add_new_class:
                addNewClass();
                break;
            case R.id.delete_all_classes:
                deleteAllClasses();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    //===========================================

    private void setAdapter() {
        classAdapter = new ClassesAdapter(classList, this);
        classContainer.setAdapter(classAdapter);

        GridLayoutManager layoutManager = new GridLayoutManager(this, 2);
        classContainer.setLayoutManager(layoutManager);
    }

    private void getClassList() {
        SQLiteDatabase readable = databaseHandler.getReadableDatabase();
        String getCommand = "SELECT * FROM " + "\"" + organisationName + "\"";
        @SuppressLint("Recycle") Cursor cursor = readable.rawQuery(getCommand, null);

        if (cursor.moveToFirst())
        {
            do {
                ClassesModel model = new ClassesModel();
                model.setId(cursor.getInt(0));
                model.setClassName(cursor.getString(1));
                model.setClassAttendancePercentage(cursor.getInt(2));
                model.setRequiredAttendance(cursor.getInt(3));
                model.setClassHistory(cursor.getString(4));
                model.setClassCounter();
                classList.add(model);
                classAdapter.notifyItemInserted(classList.indexOf(model));
            }while (cursor.moveToNext());
        }

        cursor.close();
        readable.close();

    }

    //  ===============  ADD CLASS METHODS  =============== //
    public void addNewClass(){
        AddDialogBox addDialogBox = new AddDialogBox();
        addDialogBox.show(getSupportFragmentManager(), "add dialog");
        AddDialogBox.overallAttendance = overallRequiredAttendance;
    }

    private void deleteAllClasses(){
        new AlertDialog.Builder(OrganisationActivity.this)
                .setMessage("This will delete all the Classes in the organisation. Continue?")
                .setPositiveButton("Yes", (dialog, which) -> {
                    for(ClassesModel model: classList)
                        databaseHandler.deleteClass(organisationName, String.valueOf(model.getId()));
                    classList.clear();
                    classAdapter.notifyDataSetChanged();
                    Toast.makeText(OrganisationActivity.this, "Deleted All Classes", Toast.LENGTH_LONG).show();
                    refreshProgress();
                })
                .setNegativeButton("No", (dialog, which) -> dialog.dismiss())
                .show();
    }

    @Override
    public void addSubmitDetails(EditText newNameText, EditText newTargetText) {
        if (isDataValid(newNameText, newTargetText)){
            ContentValues values = new ContentValues();
            ClassesModel model = new ClassesModel(newNameText.getText().toString(), Integer.parseInt(newTargetText.getText().toString()));
            values.put(Params.NAME, model.getClassName());
            values.put(Params.TARGET, model.getRequiredAttendance());
            values.put(Params.ATTENDANCE, model.getClassAttendancePercentage());
            values.put(Params.HISTORY, model.getClassHistory());

            int classId = databaseHandler.addNewClass(organisationName, values);
            model.setId(classId);

            classList.add(model);
            classAdapter.notifyItemInserted(classList.indexOf(model));

            refreshProgress();
        }
    }

    //=================================================================================//


    // Click Listeners

    private void setClickListeners(){
        classAdapter.setOnItemClickListener(new ClassesAdapter.OnItemClickListener() {
            @Override
            public void onHistoryClick(int position) {
                Intent intent = new Intent(OrganisationActivity.this, CalendarActivity.class);
                String[] dataArray = {organisationName, String.valueOf(classList.get(position).getId()), classList.get(position).getClassName()};
                intent.putExtra(Params.CLASS_DATA_ARRAY, dataArray);
                startActivity(intent);
            }

            @Override
            public void onDeleteClick(int position) {
                deleteClass(position);
            }

            @Override
            public void onEditClick(int position) {
                focusedClass = classList.get(position);
                editClass();
            }

            @Override
            public void onMarkClick(int position) {
                markAttendance(position);
            }

            @Override
            public void onDownloadClick(int position) {
                focusedClass = classList.get(position);
                downloadClassAttendance();
            }
        });

    }
    
    //========== DOWNLOAD ATTENDANCE ============//

    private boolean hasStoragePermission(){
        return ContextCompat.checkSelfPermission(OrganisationActivity.this, Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED
                && ContextCompat.checkSelfPermission(OrganisationActivity.this, Manifest.permission.READ_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED;
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (requestCode == STORAGE_PERMISSION_TOKEN){
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED)
                downloadClassAttendance();
            else
                Toast.makeText(this, "Permission Denied", Toast.LENGTH_SHORT).show();
        }
    }

    private void downloadClassAttendance(){
        if (hasStoragePermission()){
            ClassDownloadManager classDownloadManager = new ClassDownloadManager(focusedClass, organisationName);
            setFileModelData();
            switch (classDownloadManager.downloadAttendance()){
                case ClassDownloadManager.DOWNLOAD_FAILED:
                    Toast.makeText(OrganisationActivity.this, "Download Failed", Toast.LENGTH_SHORT).show();
                    break;
                case ClassDownloadManager.DOWNLOAD_SUCCESSFUL:
                    Toast.makeText(OrganisationActivity.this, "File Saved as " + classDownloadManager.getFilePath(), Toast.LENGTH_LONG).show();
                    break;
            }
        }
        else{
            ActivityCompat.requestPermissions(OrganisationActivity.this, new String[]{Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.WRITE_EXTERNAL_STORAGE}, STORAGE_PERMISSION_TOKEN);
        }
    }
    //=======================================================================//
    // Button methods

    private void markAttendance(int position) {
        MarkDialogBox markDialogBox = new MarkDialogBox();
        markDialogBox.show(getSupportFragmentManager(), "mark attendance");
        focusedClass = classList.get(position);
    }

    private void editClass() {
        EditDialogBox editDialogBox = new EditDialogBox();
        editDialogBox.show(getSupportFragmentManager(), "edit dialog");
        EditDialogBox.name = focusedClass.getClassName();
        EditDialogBox.target = focusedClass.getRequiredAttendance();
    }

    private void deleteClass(int position){
        new AlertDialog.Builder(OrganisationActivity.this)
                    .setMessage("Are you sure you want to delete " + classList.get(position).getClassName() + " ?")
                .setPositiveButton("Yes", (dialog, which) -> {
                    databaseHandler.deleteClass(organisationName, String.valueOf(classList.get(position).getId()));
                    classAdapter.notifyItemRemoved(position);
                    Toast.makeText(OrganisationActivity.this, "Deleted " + classList.get(position).getClassName(), Toast.LENGTH_LONG).show();
                    classList.remove(position);
                    refreshProgress();
                })
                .setNegativeButton("No", (dialog, which) -> dialog.dismiss())
                .show();
    }
    //==================================================================//
    @SuppressLint("SetTextI18n")
    private void initializeOrganisation() {
        ProgressBar requiredOverallProgress = findViewById(R.id.overall_required_attendance);
        overallProgress = findViewById(R.id.overall_attendance);
        final String getCommand =  "SELECT * FROM " + Params.ORGANISATIONS + " WHERE " + "s_no" + "=" + organisationId;
        SQLiteDatabase rootReadable = databaseHandler.getReadableDatabase();
        Cursor cursor = rootReadable.rawQuery(getCommand, null);
        cursor.moveToFirst();
        organisationName = cursor.getString(1);
        overallAttendance = cursor.getInt(2);
        overallRequiredAttendance = cursor.getInt(3);

        requiredOverallProgress.setProgress(overallRequiredAttendance);

        cursor.close();
        rootReadable.close();
    }

    @Override
    public void setCounters(TextView presentCounter, TextView absentCounter) {
        int presentCount, absentCount;
        JSONObject history;
        try {
            history = new JSONObject(focusedClass.getClassHistory());
        } catch (JSONException e) {
            return;
        }
        // INITIALIZING INITIAL DATE HISTORY
        JSONArray dateHistory;
        try {
            dateHistory = (JSONArray) history.get(getCurrentDate());
        } catch (JSONException e) {
            dateHistory = new JSONArray();
            try {
                dateHistory.put(0, 0);
                dateHistory.put(1, 0);
            } catch (JSONException ignored) {
            }
        }
        try {
            presentCount = dateHistory.getInt(0);
            absentCount = dateHistory.getInt(1);
        } catch (JSONException e) {
            Toast.makeText(this, "Some Error Occurred", Toast.LENGTH_SHORT).show();
            return;
        }
        presentCounter.setText(String.valueOf(presentCount));
        absentCounter.setText(String.valueOf(absentCount));
    }

    @Override
    public void submitAttendance(int presentCount, int absentCount) {

        int[] dateHistory = new int[2];
        dateHistory[0] = (presentCount);
        dateHistory[1] = (absentCount);

        databaseHandler.markAttendance(organisationName, focusedClass, getCurrentDate(), dateHistory);
        classAdapter.notifyItemChanged(classList.indexOf(focusedClass));
        refreshProgress();

        Toast.makeText(this, "Attendance Marked", Toast.LENGTH_SHORT).show();
        focusedClass = null;
    }

    @Override
    public void submitDetails(EditText newNameText, EditText newTargetText) {
        if (isDataValid(newNameText, newTargetText))
        {
            ContentValues values = new ContentValues();
            values.put(Params.NAME, newNameText.getText().toString());
            values.put(Params.TARGET, Integer.parseInt(newTargetText.getText().toString()));

            databaseHandler.updateClass(organisationName, values, String.valueOf(focusedClass.getId()));

            focusedClass.setClassName(newNameText.getText().toString());
            focusedClass.setRequiredAttendance(Integer.parseInt(newTargetText.getText().toString()));
            classAdapter.notifyItemChanged(classList.indexOf(focusedClass));
            Toast.makeText(this, "Updated Successfully", Toast.LENGTH_LONG).show();
            focusedClass = null;
        }
    }

    private boolean isDataValid(EditText newNameText, EditText newTargetText) {
        String name;
        int target;
        try {
            name = newNameText.getText().toString();
        } catch (Exception e) {
            name = "";
        }
        try {
            target = Integer.parseInt(newTargetText.getText().toString());
        } catch (NumberFormatException e) {
            target = 101;
        }
        if (name.isEmpty()) {
            newNameText.setError("Please Enter a valid name");
            return false;
        }
        if (name.length()>30){
            newNameText.setError("The length of the name should be less than or equal to 30");
            return false;
        }
        if (target>100 || target<0){
            newTargetText.setError("Enter a valid number from 0 to 100");
            return false;
        }
        return true;
    }

    @SuppressLint("SetTextI18n")
    private void refreshProgress() {

        ImageView flag = findViewById(R.id.target_flag);
        overallAttendance = databaseHandler.refreshOverAttendance(organisationName);

        ((TextView) findViewById(R.id.overall_percentage_counter)).setText(overallAttendance + "%");

        overallProgress.setProgress(0);

        if (overallAttendance >= overallRequiredAttendance) {
            overallProgress.setProgressDrawable(ContextCompat.getDrawable(OrganisationActivity.this, R.drawable.attendance_progress));
            flag.setImageTintList(ColorStateList.valueOf(Color.parseColor("#00CF60")));
        }
        else if (overallAttendance > overallRequiredAttendance * 0.75f){
            overallProgress.setProgressDrawable(ContextCompat.getDrawable(OrganisationActivity.this, R.drawable.attendance_progress_low));
            flag.setImageTintList(ColorStateList.valueOf(Color.parseColor("#F56600")));
        }
        else{
            overallProgress.setProgressDrawable(ContextCompat.getDrawable(OrganisationActivity.this, R.drawable.attendance_progress_danger));
            flag.setImageTintList(ColorStateList.valueOf(Color.parseColor("#FF073A")));
        }

        if (overallAttendance>0)
            overallProgress.setProgress(overallAttendance);
    }

    private String getCurrentDate(){
        Calendar calendar = Calendar.getInstance();
        return getFormattedDate(calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH), calendar.get(Calendar.DATE));
    }

    private String getFormattedDate(int year, int month, int date){
        String formattedDate = (date >= 10)? String.valueOf(date): "0" + date;
        String formattedMonth = (month >= 10)? String.valueOf(month): "0" + month;
        return year + formattedMonth + formattedDate;
    }

    private void setFileModelData() {
        FileDataModel.fonts = new Typeface[]{
                ResourcesCompat.getFont(OrganisationActivity.this, R.font.halant),
                ResourcesCompat.getFont(OrganisationActivity.this, R.font.halant_medium),
                ResourcesCompat.getFont(OrganisationActivity.this, R.font.halant_semibold),
                ResourcesCompat.getFont(OrganisationActivity.this, R.font.poly),
                ResourcesCompat.getFont(OrganisationActivity.this, R.font.adamina)
        };

        FileDataModel.logo = Bitmap.createScaledBitmap(BitmapFactory.decodeResource(getResources(), R.drawable.logo), 110, 110, false);
        FileDataModel.table = Bitmap.createScaledBitmap(BitmapFactory.decodeResource(getResources(), R.drawable.attendance_table), 700, 1000, false);
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        startActivity(new Intent(OrganisationActivity.this, MainActivity.class));
    }

}

File: app/src/main/java/com/goel/attendancetracker/activities/OrganisationActivity.kt
package com.goel.attendancetracker.activities

import android.Manifest
import android.annotation.SuppressLint
import android.content.ContentValues
import android.content.Intent
import android.content.pm.PackageManager
import android.content.res.ColorStateList
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Color
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.core.content.res.ResourcesCompat
import androidx.recyclerview.widget.GridLayoutManager
import com.goel.attendancetracker.R
import com.goel.attendancetracker.adapters.ClassesAdapter
import com.goel.attendancetracker.databinding.ActivityOrganisationBinding
import com.goel.attendancetracker.dialogboxes.AddDialogBox
import com.goel.attendancetracker.dialogboxes.AddDialogBox.AddDialogListener
import com.goel.attendancetracker.dialogboxes.EditDialogBox
import com.goel.attendancetracker.dialogboxes.EditDialogBox.EditDialogListener
import com.goel.attendancetracker.dialogboxes.MarkDialogBox
import com.goel.attendancetracker.dialogboxes.MarkDialogBox.MarkAttendanceListener
import com.goel.attendancetracker.downloadmanager.ClassDownloadManager
import com.goel.attendancetracker.downloadmanager.FileDataModel
import com.goel.attendancetracker.models.ClassesModel
import com.goel.attendancetracker.utils.Constants
import com.goel.attendancetracker.utils.database.DatabaseHandler
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.lang.NullPointerException
import java.util.*

class OrganisationActivity : AppCompatActivity(), EditDialogListener, AddDialogListener,
    MarkAttendanceListener, ClassesAdapter.OnItemClickListener {

    private lateinit var binding: ActivityOrganisationBinding
    private lateinit var organisationName: String
    private lateinit var organisationId: String
    private lateinit var databaseHandler: DatabaseHandler
    private var overallAttendance = 0
    private var overallRequiredAttendance = 0
    private lateinit var classAdapter: ClassesAdapter
    private lateinit var classList: ArrayList<ClassesModel>
    private var focusedClass: ClassesModel? = null

    @SuppressLint("SetTextI18n")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityOrganisationBinding.inflate(layoutInflater)
        setContentView(binding.root)
        binding.empty.visibility = View.GONE
        organisationId = Constants.OPEN_ORG!!
        classList = ArrayList()
        databaseHandler = DatabaseHandler(this)
        initializeOrganisation()
        supportActionBar?.title = organisationName
        setAdapter()
        getClassList()
        refreshProgress()
        binding.targetValue.text = "$overallRequiredAttendance%"
    }

    //-- Context Menu
    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.organisation_toolbar_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.add_new_class -> addNewClass()
            R.id.delete_all_classes -> deleteAllClasses()
        }
        return super.onOptionsItemSelected(item)
    }

    //===========================================
    private fun setAdapter() {
        classAdapter = ClassesAdapter(classList, this, this)
        binding.classGrid.adapter = classAdapter
        val layoutManager = GridLayoutManager(this, 2)
        binding.classGrid.layoutManager = layoutManager
    }

    private fun getClassList() {
        val readable = databaseHandler.readableDatabase
        val getCommand = "SELECT * FROM \"$organisationName\""
        @SuppressLint("Recycle") val cursor = readable.rawQuery(getCommand, null)
        if (cursor.moveToFirst()) {
            do {
                val model = ClassesModel()
                model.id = cursor.getInt(0)
                model.name = cursor.getString(1)
                model.attendance = cursor.getInt(2)
                model.target = cursor.getInt(3)
                model.classHistory = cursor.getString(4)
                model.setClassCounter()
                classList.add(model)
                classAdapter.notifyItemInserted(classList.indexOf(model))
            } while (cursor.moveToNext())
        }
        checkForEmpty()
        cursor.close()
        readable.close()
    }

    //  ===============  ADD CLASS METHODS  =============== //
    private fun addNewClass() {
        val addDialogBox = AddDialogBox(this)
        addDialogBox.show(supportFragmentManager, "add dialog")
        AddDialogBox.overallAttendance = overallRequiredAttendance
    }

    private fun deleteAllClasses() {
        AlertDialog.Builder(this)
            .setMessage(getString(R.string.delete_all_prompt))
            .setPositiveButton(R.string.confirm) { _, _ ->
                for (model in classList) databaseHandler.deleteClass(
                    organisationName,
                    model.id.toString()
                )
                classList.clear()
                classAdapter.notifyDataSetChanged()
                checkForEmpty()
                Toast.makeText(this, R.string.deleted_all_classes, Toast.LENGTH_LONG)
                    .show()
                refreshProgress()
            }
            .setNegativeButton(R.string.cancel) { dialog, _ -> dialog.dismiss() }
            .show()
    }

    private fun checkForEmpty() {
        binding.empty.visibility = if (classList.isEmpty())
            View.VISIBLE
        else
            View.GONE
    }

    override fun addSubmitDetails(newNameText: EditText, newTargetText: EditText) {
        val nameValidity = Constants.getNameValidity(newNameText.text.toString())
        val targetValidity = Constants.getTargetValidity(newTargetText.text.toString())

        if (nameValidity.containsKey(false)) {
            Toast.makeText(this, nameValidity[false], Toast.LENGTH_SHORT).show()
            return
        }
        if (targetValidity.containsKey(false)) {
            Toast.makeText(this, targetValidity[false], Toast.LENGTH_SHORT).show()
            return
        }
        val className: String
        val classTarget = try {
            className = nameValidity[true]!!
            targetValidity[true]!!.toInt()
        } catch (e: NullPointerException) {
            return
        }
        val values = ContentValues()
        val model = ClassesModel(name = className, target = classTarget)
        values.put(Constants.NAME, model.name)
        values.put(Constants.TARGET, model.target)
        values.put(Constants.ATTENDANCE, model.attendance)
        values.put(Constants.HISTORY, model.classHistory)
        val classId = databaseHandler.addNewClass(organisationName, values)
        model.id = classId
        classList.add(model)
        classAdapter.notifyItemInserted(classList.indexOf(model))
        refreshProgress()
        checkForEmpty()
    }

    //========== DOWNLOAD ATTENDANCE ============//
    private fun hasStoragePermission(): Boolean {
        return (ContextCompat.checkSelfPermission(
            this, Manifest.permission.WRITE_EXTERNAL_STORAGE
        ) == PackageManager.PERMISSION_GRANTED
                && ContextCompat.checkSelfPermission(
            this, Manifest.permission.READ_EXTERNAL_STORAGE
        ) == PackageManager.PERMISSION_GRANTED)
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == STORAGE_PERMISSION_TOKEN) {
            if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED)
                downloadClassAttendance()
            else
                Toast.makeText(this, R.string.permission_denied, Toast.LENGTH_SHORT).show()
        }
    }

    private fun downloadClassAttendance() {
        if (hasStoragePermission()) {
            val classDownloadManager = ClassDownloadManager(focusedClass!!, organisationName)
            setFileModelData()
            when (classDownloadManager.downloadAttendance()) {
                ClassDownloadManager.DOWNLOAD_FAILED -> Toast.makeText(
                    this,
                    R.string.download_failed,
                    Toast.LENGTH_SHORT
                ).show()
                ClassDownloadManager.DOWNLOAD_SUCCESSFUL -> Toast.makeText(
                    this,
                    getString(R.string.file_saved, classDownloadManager.filePath),
                    Toast.LENGTH_LONG
                ).show()
            }
        } else {
            ActivityCompat.requestPermissions(
                this,
                arrayOf(
                    Manifest.permission.READ_EXTERNAL_STORAGE,
                    Manifest.permission.WRITE_EXTERNAL_STORAGE
                ),
                STORAGE_PERMISSION_TOKEN
            )
        }
    }

    //=======================================================================//
    // Button methods
    private fun markAttendance(position: Int) {
        val markDialogBox = MarkDialogBox(this)
        markDialogBox.show(supportFragmentManager, "mark attendance")
        focusedClass = classList[position]
    }

    private fun editClass() {
        val editDialogBox = EditDialogBox(this)
        editDialogBox.show(supportFragmentManager, "edit dialog")
        EditDialogBox.name = focusedClass!!.name
        EditDialogBox.target = focusedClass!!.target
    }

    private fun deleteClass(position: Int) {
        val classId = classList[position].id.toString()
        val className = classList[position].name
        databaseHandler.deleteClass(organisationName, classId)
        classAdapter.notifyItemRemoved(position)
        Toast.makeText(
            this,
            getString(R.string.deleted, className),
            Toast.LENGTH_LONG
        ).show()
        classList.removeAt(position)
        refreshProgress()
        checkForEmpty()
    }

    //==================================================================//
    private fun initializeOrganisation() {
        val requiredOverallProgress = binding.overallRequiredAttendance
        val getCommand = "SELECT * FROM ${Constants.ORGANISATIONS} WHERE s_no=$organisationId"
        val rootReadable = databaseHandler.readableDatabase
        val cursor = rootReadable.rawQuery(getCommand, null)
        cursor.moveToFirst()
        organisationName = cursor.getString(1)
        overallAttendance = cursor.getInt(2)
        overallRequiredAttendance = cursor.getInt(3)
        requiredOverallProgress.progress = overallRequiredAttendance
        cursor.close()
        rootReadable.close()
    }

    override fun setCounters(presentCounter: TextView, absentCounter: TextView) {
        val presentCount: Int
        val absentCount: Int
        val history: JSONObject = try {
            JSONObject(focusedClass!!.classHistory)
        } catch (e: JSONException) {
            return
        }
        // INITIALIZING INITIAL DATE HISTORY
        var dateHistory: JSONArray
        try {
            dateHistory = history[currentDate] as JSONArray
        } catch (e: JSONException) {
            dateHistory = JSONArray()
            try {
                dateHistory.put(0, 0)
                dateHistory.put(1, 0)
            } catch (ignored: JSONException) {
            }
        }
        try {
            presentCount = dateHistory.getInt(0)
            absentCount = dateHistory.getInt(1)
        } catch (e: JSONException) {
            Toast.makeText(this, R.string.error_message, Toast.LENGTH_SHORT).show()
            return
        }
        presentCounter.text = presentCount.toString()
        absentCounter.text = absentCount.toString()
    }

    override fun submitAttendance(presentCount: Int, absentCount: Int) {
        val dateHistory = IntArray(2)
        dateHistory[0] = presentCount
        dateHistory[1] = absentCount
        focusedClass?.let {
            databaseHandler.markAttendance(organisationName, it, currentDate, dateHistory)
        }
        classAdapter.notifyItemChanged(classList.indexOf(focusedClass))
        refreshProgress()
        Toast.makeText(this, R.string.attendance_marked, Toast.LENGTH_SHORT).show()
        focusedClass = null
    }

    override fun submitDetails(newNameText: EditText, newTargetText: EditText) {
        val nameValidity = Constants.getNameValidity(newNameText.text.toString())
        val targetValidity = Constants.getTargetValidity(newTargetText.text.toString())

        if (nameValidity.containsKey(false)) {
            Toast.makeText(this, nameValidity[false], Toast.LENGTH_SHORT).show()
            return
        }
        if (targetValidity.containsKey(false)) {
            Toast.makeText(this, targetValidity[false], Toast.LENGTH_SHORT).show()
            return
        }
        val className: String
        val classTarget = try {
            className = nameValidity[true]!!
            targetValidity[true]!!.toInt()
        } catch (e: NullPointerException) {
            return
        }
        val values = ContentValues()
        values.put(Constants.NAME, className)
        values.put(Constants.TARGET, classTarget)
        databaseHandler.updateClass(organisationName, values, focusedClass!!.id.toString())
        focusedClass!!.name = className
        focusedClass!!.target = classTarget
        classAdapter.notifyItemChanged(classList.indexOf(focusedClass))
        Toast.makeText(this, R.string.updated, Toast.LENGTH_LONG).show()
        focusedClass = null
    }

    @SuppressLint("SetTextI18n")
    private fun refreshProgress() {
        val overallProgress = binding.overallAttendance
        val flag = binding.targetFlag
        overallAttendance = databaseHandler.refreshOverAttendance(organisationName)
        binding.overallPercentageCounter.text = "$overallAttendance%"
        overallProgress.progress = 0
        when {
            overallAttendance >= overallRequiredAttendance -> {
                overallProgress.progressDrawable =
                    ContextCompat.getDrawable(this, R.drawable.attendance_progress)
                flag.imageTintList = ColorStateList.valueOf(Color.parseColor("#00CF60"))
            }
            overallAttendance > overallRequiredAttendance * 0.75f -> {
                overallProgress.progressDrawable = ContextCompat.getDrawable(
                    this,
                    R.drawable.attendance_progress_low
                )
                flag.imageTintList = ColorStateList.valueOf(Color.parseColor("#F56600"))
            }
            else -> {
                overallProgress.progressDrawable = ContextCompat.getDrawable(
                    this,
                    R.drawable.attendance_progress_danger
                )
                flag.imageTintList = ColorStateList.valueOf(Color.parseColor("#FF073A"))
            }
        }
        if (overallAttendance > 0) overallProgress.progress = overallAttendance
    }

    private val currentDate: String
        get() {
            val calendar = Calendar.getInstance()
            return getFormattedDate(
                calendar[Calendar.YEAR],
                calendar[Calendar.MONTH],
                calendar[Calendar.DATE]
            )
        }

    private fun getFormattedDate(year: Int, month: Int, date: Int): String {
        val formattedDate = if (date >= 10) date.toString() else "0$date"
        val formattedMonth = if (month >= 10) month.toString() else "0$month"
        return year.toString() + formattedMonth + formattedDate
    }

    private fun setFileModelData() {
        FileDataModel.fonts = arrayOf(
            ResourcesCompat.getFont(this, R.font.halant),
            ResourcesCompat.getFont(this, R.font.halant_medium),
            ResourcesCompat.getFont(this, R.font.halant_semibold),
            ResourcesCompat.getFont(this, R.font.poly),
            ResourcesCompat.getFont(this, R.font.adamina)
        )
        FileDataModel.logo = Bitmap.createScaledBitmap(
            BitmapFactory.decodeResource(resources, R.drawable.logo),
            110,
            110,
            false
        )
        FileDataModel.table = Bitmap.createScaledBitmap(
            BitmapFactory.decodeResource(
                resources,
                R.drawable.attendance_table
            ), 700, 1000, false
        )
    }

    override fun onHistoryClick(position: Int) {
        CalendarActivity.model = classList[position]
        CalendarActivity.organisationName = organisationName
        val intent = Intent(this, CalendarActivity::class.java)
        startActivity(intent)
    }

    override fun onDeleteClick(position: Int) {
        val name = classList[position].name
        AlertDialog.Builder(this)
            .setMessage(getString(R.string.delete_class_prompt, name))
            .setPositiveButton(R.string.confirm) { _, _ ->
                deleteClass(position)
            }
            .setNegativeButton(R.string.cancel) { dialog, _ -> dialog.dismiss() }
            .show()
    }

    override fun onEditClick(position: Int) {
        focusedClass = classList[position]
        editClass()
    }

    override fun onMarkClick(position: Int) {
        markAttendance(position)
    }

    override fun onDownloadClick(position: Int) {
        focusedClass = classList[position]
        downloadClassAttendance()
    }

    override fun onBackPressed() {
        super.onBackPressed()
        startActivity(Intent(this, MainActivity::class.java))
    }

    companion object {
        private const val STORAGE_PERMISSION_TOKEN = 1
    }
}